pkg/daemon/controller/controller.go

```diff
@@ -0,0 +1,311 @@
+	if config.Description != expectedDescription {
+		config.Description = expectedDescription
+		updateRequired = true
+	}
+
+	if !reflect.DeepEqual(config.Dependencies, expected.Dependencies) {
+		config.Dependencies = expected.Dependencies
+		updateRequired = true
+	}
+
+	if updateRequired {
+		// Always stop the service before updating its config, just to be safe
+		if _, err := service.Control(svc.Stop); err != nil {
+			return err
+		}
+		err = service.UpdateConfig(config)
+		if err != nil {
+			return errors.Wrap(err, "error updating service config")
+		}
+	}
+
+	// always ensure service is started
+	return service.Start()
+}
+
+// expectedServiceCommand returns the full command that the given service should run with
+func (sc *ServiceController) expectedServiceCommand(expected servicescm.Service) (string, error) {
+	var nodeVars, psVars map[string]string
+	var err error
+	if len(expected.NodeVariablesInCommand) > 0 {
+		nodeVars, err = sc.resolveNodeVariables(expected)
+		if err != nil {
+			return "", err
+		}
+	}
+	if len(expected.PowershellVariablesInCommand) > 0 {
+		psVars, err = sc.resolvePowershellVariables(expected)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	expectedCmd := expected.Command
+	for key, value := range nodeVars {
+		expectedCmd = strings.ReplaceAll(expectedCmd, key, value)
+	}
+	for key, value := range psVars {
+		expectedCmd = strings.ReplaceAll(expectedCmd, key, value)
+	}
+	return expectedCmd, nil
+}
+
+// resolveNodeVariables returns a map, with the keys being each variable, and the value being the string to replace the
+// variable with
+func (sc *ServiceController) resolveNodeVariables(svc servicescm.Service) (map[string]string, error) {
+	vars := make(map[string]string)
+	var node core.Node
+	err := sc.client.Get(sc.ctx, client.ObjectKey{Name: sc.nodeName}, &node)
+	if err != nil {
+		return nil, err
+	}
+	for _, nodeVar := range svc.NodeVariablesInCommand {
+		nodeParser := jsonpath.New("nodeParser")
+		if err := nodeParser.Parse(nodeVar.NodeObjectJsonPath); err != nil {
+			return nil, err
+		}
+		values, err := nodeParser.FindResults(node)
+		if err != nil {
+			return nil, err
+		}
+		if len(values) == 0 {
+			return nil, errors.Wrapf(err, "expected node value %s missing", nodeVar.NodeObjectJsonPath)
+		}
+		if len(values) > 1 {
+			return nil, errors.Wrapf(err, "jsonpath %s returned too many results", nodeVar.NodeObjectJsonPath)
+		}
+		if len(values[0]) != 1 || values[0][0].Kind() != reflect.String {
+			return nil, errors.Wrapf(err, "unexpected value type for %s", nodeVar.NodeObjectJsonPath)
+		}
+		vars[nodeVar.Name] = values[0][0].String()
+	}
+	return vars, nil
+}
+
+// resolvePowershellVariables returns a map, with the keys being each variable, and the value being the string to replace the
+// variable with
+func (sc *ServiceController) resolvePowershellVariables(svc servicescm.Service) (map[string]string, error) {
+	// TODO: Implement this function
```

Comment:
Will this be further implemented out in future controller PRs or can this TODO be removed?